pkgs/dart_model/lib/src/json.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2024, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:collection/collection.dart';
+
+import 'json_changes.dart';
+
+/// JSON data.
```

Comment:
This is a crucial part of the design but still a bit fuzzy, let me try writing it out once to see if that helps :)

This class is to encapsulate json data so that we can choose between the default Map based representation and a fast byte buffer representation. It doesn't know or care about the structure of the data, but for performance it will care about how it's instantiated. It also cares about serialization.

Meaning will be attached to the data via extension types over maps/lists in the data; those maps and lists might be standard types or byte buffer backed special types, depending on the representation being used.

The fuzzy parts: I'm not sure yet about mutability/copying and how that needs to be set up to support convenient and fast operation in both modes.

My current idea to flesh this out is by adding a first simple schema, generating some extension types, then we can start to see how the whole thing ties together.

Reply:
Okay, I think I made some progress :)

Renamed this class to `JsonData`, it's a blob of immutable + serializable JSON data.

You access it via `asMap` which returns a read-only `JsonMap`.

We don't actually need == / hashCode, removed those.

This fits exactly with the `JsonBuffer` exploration, where there is one class holding the byte buffer and thin classes provide views into it. It also clears up the question of immutability :)